Python/pywarpx/_libwarpx.py

```diff
@@ -30,112 +30,294 @@
 
-# this is a plain C/C++ shared library, not a Python module
-if os.name == 'nt':
-    mod_ext = "dll"
-else:
-    mod_ext = "so"
-libname = "libwarpx.{0}.{1}".format(geometry_dim, mod_ext)
-
-try:
-    libwarpx = ctypes.CDLL(os.path.join(_get_package_root(), libname))
-except OSError as e:
-    value = e.args[0]
-    if f'{libname}: cannot open shared object file: No such file or directory' in value:
-        raise Exception(f'"{libname}" was not installed. Installation instructions can be found here https://warpx.readthedocs.io/en/latest/install/users.html') from e
-    else:
-        print("Failed to load the libwarpx shared object library")
-        raise
+class LibWarpX():
+    def __init__(self):
+        self.__dict__['library_loaded'] = False
 
-# track whether libwarpx has been initialized
-libwarpx.initialized = False
+        # track whether amrex and warpx have been initialized
+        self.__dict__['initialized'] = False
 
-# WarpX can be compiled using either double or float
-libwarpx.warpx_Real_size.restype = ctypes.c_int
-libwarpx.warpx_ParticleReal_size.restype = ctypes.c_int
 
-_Real_size = libwarpx.warpx_Real_size()
-_ParticleReal_size = libwarpx.warpx_ParticleReal_size()
+    def __getattr__(self, attribute):
+        if not self.library_loaded:
+            self.load_library()
+        try:
+            return self.__dict__[attribute]
+        except KeyError:
+            return getattr(self.libwarpx, attribute)
+
+    def get_package_root(self):
+        '''
+        Get the path to the installation location (where libwarpx.so would be installed).
+        '''
+        cur = os.path.abspath(__file__)
+        while True:
+            name = os.path.basename(cur)
+            if name == 'pywarpx':
+                return cur
+            elif not name:
+                return ''
+            cur = os.path.dirname(cur)
+
+    def load_library(self):
+        self.library_loaded = True
+
+        # --- Use geometry to determine whether to import the 1D, 2D, 3D or RZ version.
+        # --- The geometry must be setup before the lib warpx shared object can be loaded.
+        try:
+            _prob_lo = geometry.prob_lo
+            _coord_sys = geometry.coord_sys
+        except AttributeError:
+            raise Exception('The geometry must be setup before the WarpX shared object can be accessesd')
+
+        if _coord_sys == 0:
+            self.geometry_dim = '%dd'%len(_prob_lo)
+        elif _coord_sys == 1:
+            self.geometry_dim = 'rz'
+        else:
+            raise Exception('Undefined coordinate system %d'%_coord_sys)
 
-if _Real_size == 8:
-    c_real = ctypes.c_double
-    _numpy_real_dtype = 'f8'
-else:
-    c_real = ctypes.c_float
-    _numpy_real_dtype = 'f4'
+        # this is a plain C/C++ shared library, not a Python module
+        if os.name == 'nt':
+            mod_ext = "dll"
+        else:
+            mod_ext = "so"
+        libname = "libwarpx.{0}.{1}".format(self.geometry_dim, mod_ext)
```

Comment:
```suggestion
        libname = f"libwarpx.{self.geometry_dim}.{mod_ext}"
```
Should f strings be used here for consistency since other new changes also use f strings?

Reply:
Fixed